add 'lean' mode for Lean Theorem Prover files

soonhokong opened the pull request:
We're using it at https://leanprover.github.io/tutorial

nightwing commented on lib/ace/mode/lean_highlight_rules.js:
```diff
@@ -0,0 +1,224 @@
+                regex : "\\s+"
+            }
+        ],
+        "comment" : [
```
This is better written as `[ {token: "comment", regex: "-/", next: "start"}, {defaultToken: "comment"} ]`
Comment spanning whole line trick was necessary in old versions where defaultToken wasn't supported.

nightwing commented on lib/ace/mode/lean_highlight_rules.js:
```diff
@@ -0,0 +1,224 @@
+                next : "start"
+            }, {
+                token : "string",
+                regex : '.+'
```
same here

nightwing commented on lib/ace/ext/modelist.js:
```diff
@@ -99,6 +99,7 @@ var supportedModes = {
     JSX:         ["jsx"],
     Julia:       ["jl"],
     LaTeX:       ["tex|latex|ltx|bib"],
+    Lean:        ["lean|hlean"],
```
Could you also add a small demo lean file to https://github.com/ajaxorg/ace/tree/master/demo/kitchen-sink/docs, just  a few lines would be fine

soonhokong commented:
@nightwing, thanks for the comments. I'll fix them.

soonhokong commented:
@nightwing, I've added things that you suggested. Could you review those commits? Thank you!

nightwing commented on demo/kitchen-sink/docs/lean.lean:
```diff
@@ -0,0 +1,462 @@
+
+/- TODO: Multiplication and one, starting with mult_right_le_one_le. -/
+
+end algebra
```
Could you please make this a _few lines_. Something closer to 10 than 500:)

soonhokong commented:
@nightwing, I just made a destructive update https://github.com/ajaxorg/ace/commit/8f49d3d89784095089bf0885268b4e87e2fbc69c. It's 10 lines long.

nightwing commented on lib/ace/mode/lean_highlight_rules.js:
```diff
@@ -0,0 +1,200 @@
+        ],
+        "comment" : [ {token: "comment", regex: "-/", next: "start"},
+                      {defaultToken: "comment"} ],
+        "qqstring" : [ {token : "string", regex : '(?:(?:\\\\.)|(?:[^"\\\\]))*?"', next : "start" },
```
This should be something like

```js
[ {
     token : "constant.language.escape",
     regex : /\\./
}, {
     token : "string",
     regex : '"',
     next : "start" 
}, {
    defaultToken: "string"
}]
```

This depends on whether strings are multiline or not, and are there special escape patterns like \uXXXX

soonhokong commented:
@nightwing, thanks for the comment. I've cleaned up many things and updated a rule for string. Please review them, esp 34d80e81cafe90ea78bdb5fcc05fad8a0076bbba.

soonhokong commented:
@nightwing, FYI, 706ee70affc739b5e6943e8745b2d09b1f6352b9 is pushed.

nightwing commented:
Pulled. 
Thank you!

soonhokong commented:
@nightwing, really appreciate your help!
